[jit] kill module_lookup

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23098 [wip][jit] serialize modules as classes
* #23154 [jit] Put all modules in the global Python CU
* #23170 [jit] Make optimize a thread_local flag
* #23109 [jit] add simple inheritance support to AST
* **#23097 [jit] kill module_lookup**
* #23033 [jit] make RHS  of assignment optional
* #23031 [jit] Parse all stmts in class defs

Instead of this module lookup thing, just wire up the submodule
hierarchy directly during the import process. In Python, construct the
mirror Python hierarchy after the fact.

gh-metadata: pytorch pytorch 23097 gh/suo/98/head

Differential Revision: [D16383329](https://our.internmc.facebook.com/intern/diff/D16383329)

suo committed:
[jit] do base module initialization lazily on import

We need this because we don't actually have a name for the base module
until after we begin the process of deserialization. Previously, we were
giving it a bogus name, like "__main__".

This PR changes ModuleLookup to be a functor that initializes the base
module with the right name when it's first looked up.

suo committed:
Update on "[jit] do base module initialization lazily on import"

[jit] do base module initialization lazily on import

We need this because we don't actually have a name for the base module
until after we begin the process of deserialization. Previously, we were
giving it a bogus name, like "__main__".

This PR changes ModuleLookup to be a functor that initializes the base
module with the right name when it's first looked up.

gh-metadata: pytorch pytorch 23097 gh/suo/98/head

suo committed:
Update on "[jit] do base module initialization lazily on import"

[jit] do base module initialization lazily on import

We need this because we don't actually have a name for the base module
until after we begin the process of deserialization. Previously, we were
giving it a bogus name, like "__main__".

This PR changes ModuleLookup to be a functor that initializes the base
module with the right name when it's first looked up.

gh-metadata: pytorch pytorch 23097 gh/suo/98/head

suo committed:
Update on "[jit] do base module initialization lazily on import"

[jit] kill module_lookup

Instead of this module lookup thing, just wire up the submodule
hierarchy directly during the import process. In Python, construct the
mirror Python hierarchy after the fact.

gh-metadata: pytorch pytorch 23097 gh/suo/98/head

zdevito reviewed (approved):
Looks good and it is simpler too.

suo committed:
Update on "[jit] kill module_lookup"

[jit] kill module_lookup

Instead of this module lookup thing, just wire up the submodule
hierarchy directly during the import process. In Python, construct the
mirror Python hierarchy after the fact.

gh-metadata: pytorch pytorch 23097 gh/suo/98/head
